### DISCUSSION TITLE
Opinions sought on a new conditional return keyword "returnon".

### DISCUSSION THREAD

**Discussion by The-Futurist** [dotnet/csharplang#5458] [Language Ideas]
I often find myself writing stuff like this:

```cs

if (!MethodA()) 
   return false;

if (!MethodB())
   return false;

if (MethodC())
   return true;

return MethodD();

```

it therefore struck me that two new keywords might be helpful here:

```cs

refalse MethodA();

refalse MethodB();

retrue MethodC();

return MethodD();
```

Obviously `retrue` returns true if the expression is true, else it continues execution and `refalse` returns false if the expression is false.

Something _like_ this can be achieved with this:

```cs

return MethodA() && MethodB() && MethodC() && MethodD();
```

This relies on "short circuit" behavior of logical expressions and the above would return `false` if _any_ were false and only return `true` if _all_ of them were true, of course each is executed in left to right order.

But the downside here is that we might (and often do) have additional steps between each of the `return` statements above and that can't easily be accommodated in a simple logical expression.

Yes the terms `retrue` and `refalse` are a bit crude but you get the idea, its not a huge saving in typing etc but is less cluttered than it is today, so has such a concept or feature ever been considered for C#?

**Comment by HaloFour** [comment]
This seems incredibly niche.  How often do you really need to shortcircuit a `bool`-returning function based on the same `bool` being returned by some expression?  What about returning `true` due to an expression returning `false`?

**Comment by CyrusNajmabadi** [comment]
I genuinely didn't realize these were different:

```c#
retrue MethodC();

return MethodD();
```

`retrue` and `return` are so visually similar (with just transposed `r` and `u`, and single different last letter) that it's a non-starter  to have the language use this.

**Comment by The-Futurist** [comment]
This is fair feedback, I just wanted to convey the problem but I agree the suggested means is not great. 

Here's what I think is much better and more general:

```cs

return TrySomethingOrOther(), false;

// other code is here.

return TryThisOrThat(), false;

// other code is here.

return TryThisAsTheLastStep();
```

This would expand the return statement from:

```cs

return <expression> ;
```
to
```cs
return <expression_1> [ , <expression_2> ] ;
```
(Where [ ] denotes an _optional_ clause);

With the meaning being _only_ return when <expression_1> is _equal_ to <expression_2> and the value to return being <expression_1>, otherwise continue execution at the next statement. This does not require a new keyword nor would it lead to backward compatibility difficulties it also isn't restricted to simply `bool` valued expressions, _any_ kind of expression can be used.

The entire idea arose from the real world, working on a code base that has a large number of mutually interacting `TryXXX` methods, where the pattern (see very first code snippet in the OP) soon becomes tiresome when handled with the language as it currently stands.

An additional variant also springs to mind:

```cs
return <expression_1> [ , <expression_2> [ , <value_1> ] ] ;
```
for example:

```cs
return GetCountRemaining(), 0, true;
```

Here a boolean `true` result is returned if the two expressions are equal.

Or how about:

```cs
return <boolean_valued_expression> [ , <value> ] ;
```

e.g.

```cs
return GetCountRemaining() == 0, true;
```
e.g.
```cs
public bool AreWeDoneYet()
{
   return GetCountRemaining() > 0 && connected == true, false;

   // do other stuff here...
}
```

The visual cue of the comma tells us right away that this particular `return` statement _may_ or _may not_ return.

Of course - at least in C# today - this might be a problem:

```cs
public bool AreWeDoneYet()
{
   return (GetCountRemaining() > 0 && connected == true, false);

   // do other stuff here...
}
```
Here, adding an "innocent" pair of parentheses changes it to a conventional return of a tuple, so for this reason we could do this:

```cs
public bool AreWeDoneYet()
{
   returnon GetCountRemaining() > 0 && connected == true, false; // new keyword `returnon` 

   // do other stuff here...
}
```
with this then becoming a syntax error:

```cs
public bool AreWeDoneYet()
{
   returnon (GetCountRemaining() > 0 && connected == true, false); // new keyword `returnon` 

   // do other stuff here...
}
```
because the syntax is:

```cs
returnon <boolean_valued_expression> , <value> ;
```

the `<value>` no longer being optional.

Finally given the way `=>` has been leveraged for properties etc recently.

```cs
public bool AreWeDoneYet()
{
   returnon GetCountRemaining() > 0 && connected == true => false; // new keyword `returnon` 

   // do other stuff here...
}
```

where the syntax is then:

```cs
returnon <boolean_valued_expression> => <value> ;
```

With this in place the initial example:

```cs
if (!MethodA()) 
   return false;

if (!MethodB())
   return false;

if (MethodC())
   return true;

return MethodD();
```

can be rewritten as:

```cs

returnon !MethodA() => false;

returnon !MethodB() => false;

returnon MethodC() => true;

return MethodD();
```

**Comment by iam3yal** [comment]
Can't you do something like this? 
```csharp
var x = MethodA();
x = x ? MethodB() : x;
x = x ? MethodN() : x;
return x ? MethodC() ? x : MethodD() : x;
```

**Comment by jnm2** [comment]
Similar conversations: https://github.com/dotnet/csharplang/discussions/325, https://github.com/dotnet/csharplang/discussions/4173, https://github.com/dotnet/csharplang/discussions/4711

**Comment by MgSam** [comment]
``` C#
if (WritingTooManyChecks())
    reassess Life();
```

**Comment by theunrepentantgeek** [comment]
I believe the C# team strongly prefers generic (widely useful) syntax options over niche (narrowly useful) ones.

Instead of syntax that only works with **bool** values (the originally suggested **refalse** and **retrue** keywords), or the later suggestion (the one combining `return <expression_1> [ , <expression_2> ] ;` with some arcane rules), I'd suggest looking to existing keywords and syntax structures for a potential solution.

One approach would be to use the keyword **if** to specify a predicate for the **return** statement. 

The original example would then become:

``` csharp
return false if !MethodA()
return false if !MethodB()
return true if MethodC()
return MethodD();
```

The advantage here is that there's no restriction on the return type, and no additional complex rules - someone who is already familiar with any pre-existing version of C# (or indeed almost any language with similar syntax) would likely intuit the actual behaviour.

Another approach would be to use the **when** keyword - which is already used in exactly this fashion for the **catch** statement:

``` csharp
return false when !MethodA()
return false when !MethodB()
return true when MethodC()
return MethodD();
```

_**However**_ ... while I think the above syntax suggestions would be useful additions to the language, I don't think any changes are actually warranted in this area. I feel as though this issue is discussing solutions looking for problems, not the other way around.

**Reply by DavidArno** [reply]
I read your code samples and thought "I can already express that as":
```c#
if (!MethodA()) return false;
if (!MethodB()) return false;
if (MethodC()) return true;
return MethodD();
```
and realised that was the OP's original code ... which suggests to me the C# already has a viable solution to the problem and so why add a new approach? So I agree with you: I don't think any changes are actually warranted in this area either.

**Reply by The-Futurist** [reply]
> 

@theunrepentantgeek 

I think:

```cs
return false when !MethodA()
```

is quite interesting, this is a good syntax for some of the reasons you mention.

This could be expanded to:

```cs
return <expression> when <predicate> ;
```

I do like this a lot, I think this addresses the problem I mentioned in my OP situations where there is a large number of opportunities to return some value as we progress down a list of statements.

**Reply by The-Futurist** [reply]
> 

@DavidArno of course there are existing ways to achieve the same ends but these statements all start with "if". In a statement that begins with "if" a huge number of consequential actions can be written, return (exiting the code) is just one of them. The eye is drawn to the "if" not the return, the intent of the statement appears to be asking a question.

In the reverse syntax it is clear that an exit from the method is the focus, the eye is drawn to the "return" clearly this is about "returning" not about asking a question.

Of course logically they are identical but the visual impact is different, one emphasizes asking a question whereas the other emphasizes returning from the method, I'd argue the additional expressiveness is valuable and the cost of implementing the change to the language is relatively small.

**Reply by HaloFour** [reply]
> The eye is drawn to the "if" not the return, the intent of the statement appears to be asking a question.

And that's intentional, the "if" is the most important part of that statement.  The "return" can only be a consequence of that "if".

IMO it's not worth adding conditional versions of any/all statements to the language.  These are all already expressable using existing idiomatic syntax.  It adds a lot of complexity for the sake of stylistic preference.

**Reply by The-Futurist** [reply]
> 

@HaloFour - Well it strikes me as being pretty _easy_ to add to the language, nothing complex as far as I can see. I certainly the feel the extra expressiveness and readability available for some developers in some situations would more than justify the small cost of adjusting the language.

There seems to be no backward compatibility issues either with:

```cs

return <expression> when <predicate> ;

```

because `return` today is either

```cs
return;
```

or

```cs
return <expression> ;
```

so adding a `when` keyword is pretty straightforward, so what do you mean when you say "a lot of complexity"? 

It's worth noting too that conditional return instructions have been present in computer instructions sets for decades.

**Reply by CyrusNajmabadi** [reply]
> @HaloFour - Well it strikes me as being pretty easy to add to the language, nothing complex as far as I can see.

Complexity goes beyond implementation side of things.  Complexity also has to be assessed in terms of how it impacts things like readability, user confusion, etc.

IMO, i'm in the position that this would not be a good addition to the language.  It inverts things and looks misleading.  `return` indicates taht the scope will be returned from.  Now that isn't hte case.  The return has become conditional, but you have to scan forward to realize that.  If it's a conditional return, then i think the least complex and more appropriate way to do that is explicitly to just have the condition first, followed by teh return.  This makes it easy to read, write and understand.  THe first thing you see is the more important part (the condition that branches) then the consequence of that branch.  As opposed to seeing something which looks like it will happen, only to find it there are conditions on it that follow after the fact.

It would be akin to having a linq query with a 'select' which was then followed by a 'where'.  The branching and conditions impact what happens, so they should ideally come first.

**Reply by CyrusNajmabadi** [reply]
> The eye is drawn to the "if" not the return, the intent of the statement appears to be asking a question.

IMO, the intent of the statement is to ask a question, and then to jump to one of two places depending on the answer to that question.

> In the reverse syntax it is clear that an exit from the method is the focus, the eye is drawn to the "return" clearly this is about "returning" not about asking a question.

Hrmm... but returning is not the focus (afaict).  Indeed, in 1/2 of hte condition results, it won't return but will instead jump to the next statement.  

>  I'd argue the additional expressiveness is valuable and the cost of implementing the change to the language is relatively small.

So this will ultimately come down to subjective interpretation.  My read of the above is that this expressiveness is not helpful here, and is actually substantively detrimental to readability (esp. given the history of the language and waht `return` has always meant).

**Reply by CyrusNajmabadi** [reply]
I agree with David here.  This feels well solved by a very idiomatic and clear construct already in the language which has served very well for the last 22 years.  THe new construct actually feels more confusing to me, and doesn't seem to save anything.  THe expressivity gain actually seems to be a net loss from my read.

**Reply by HaloFour** [reply]
Complexity includes adding multiple ways of doing things.  It's not just about being technically challenging or difficult syntactically.  This is adding a new form of an existing statement only to duplicate the functionality that can be accomplished by combining two orthogonal statements without any appreciable change in verbosity.  It also begs the question about adding such forms to any other control statement, like `throw`, `continue`, `yield`, `break`, etc.

Also, see: https://github.com/dotnet/csharplang/issues/2966

**Reply by DavidArno** [reply]
> @DavidArno of course there are existing ways to achieve the same ends but these statements all start with "if". In a statement that begins with "if" a huge number of consequential actions can be written, return (exiting the code) is just one of them. The eye is drawn to the "if" not the return, the intent of the statement appears to be asking a question.

That's a good point. Many companies enforce rules around braces and multi-lines here, so:
```c#
return false when !MethodA();
```
has to be compared to:
```c#
if (!MethodA())
{
    return false;
}
```
However, when I look at that first item, I find I mentally read it thus: "OK, so there's a when there, so if methodA is false, then we'll return". That could be because of years of conditioning in reading C/C++/Java/C# etc code for me of course, but since that's the way I process it, I like it being the if then return order in code too. Sorry, but I'm not a fan of offering the option of specifying it in the opposite order simply as an alternative way of doing things therefore.

**Reply by The-Futurist** [reply]
> 

@HaloFour "Complexity includes adding multiple ways of doing things". Well clearly your not swayed and that's fine, this was just an idea that arose as I was staring at some code I was writing that had a larger number of conditional returns than is typical.

But as for multiple ways to do things, let's not forget:

```cs
a = a + 1;
a += 1;
a++;
++a;
```

**Reply by CyrusNajmabadi** [reply]
This discussion is not around compound operators though.  It's about branching and exiting.  The rules and beliefs around one don't necessarily apply to the other. 

c# does not have the position that there can't be more than one way to do something.  The position it has though is that if there are going to be multiple ways there has to be a strong belief that the different ways are bring a lot of positives to the language that warrants the complexity that Halo is mentioning.

**Comment by osamakawish** [comment]
So I guess you want something like: 

```c#
if (!MethodA()) 
   return false;

DoStuffB();

if (!MethodB())
   return false;

DoStuffC();

if (MethodC())
   return true;

DoStuffD();

return MethodD();
```

You should be able to write a method for something like this, then combine it with the boolean output. 
Then a simple boolean expression should give you want you want:

```c#
public Func<bool> PrelimAndCondition(Action prelim, Func<bool> condition) { prelim(); return condition(); }

return MethodA() && PrelimAndCondition(DoStuffB, MethodB) && PrelimAndCondition(DoStuffC, MethodC)
    && PrelimAndCondition(DoStuffD, MethodD);
```
